Fix #85 and #45...

raulccabreu opened the pull request:
...with a endless scroll that fetchs on Rifty if "hasMoreGifs = true". The imported Gist class EndlessRecyclerScrollListener implements this behavior.

- Add a initialState method that initialize activity attributes based on initial AppState.
- Adjust some references to word "Component" (used before on FeedComponent and GifComponent) that now is only "View" following new naming of FeedView and GifView.
- Move the "package" net to inside package "data" where DataManager is the interface of the package as defined on planned architecture.
- Create the locker RecyclerTestLocker that lock based on RecyclerView's adapter size.
- Create the base class BaseTestLocker with common code between AnvilTestLocker and RecyclerTestLocker.
- Add the tests: "whenGifListIsEmpty", "whenGifListIsNotEmpty", "whenDefaultGifsAreLoaded", "whenEndlessScrollTriggered" and "whenEndlessScrollNotTriggered".
- Move some methods from GifListActivityTest to TestHelper and reuse this methods on other test classes.

raulccabreu committed:
Fix #85 and #45 with a endless scroll that fetchs on Rifty if "hasMoreGifs = true". The imported Gist class EndlessRecyclerScrollListener implements this behavior.

- Add a initialState method that initialize activity attributes based on initial AppState.
- Adjust some references to word "Component" (used before on FeedComponent and GifComponent) that now is only "View" following new naming of FeedView and GifView.
- Move the "package" net to inside package "data" where DataManager is the interface of the package as defined on planned architecture.
- Create the locker RecyclerTestLocker that lock based on RecyclerView's adapter size.
- Create the base class BaseTestLocker with common code between AnvilTestLocker and RecyclerTestLocker.
- Add the tests: "whenGifListIsEmpty", "whenGifListIsNotEmpty", "whenDefaultGifsAreLoaded", "whenEndlessScrollTriggered" and "whenEndlessScrollNotTriggered".
- Move some methods from GifListActivityTest to TestHelper and reuse this methods on other test classes.

niltonvasques commented on app/src/androidTest/java/br/com/catbag/gifreduxsample/ui/giflist/GifListActivityTest.java:
```diff
@@ -214,9 +227,29 @@ public void whenGifDownloadFailTest() {
+        mActivityTestRule.launchActivity(new Intent());
+
+        onView(withId(R.id.loading))
+                .check(matches(withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE)));
```
When the list is empty, the loading will run forever?

raulccabreu replied:
If the AppState don't change, yes. But without mocks the activity calls the load action, and the Database populates with default gifs...if no previus AppState was saved.

fpbitencourt replied:
Em breve não teremos mais esse preenchimento automático, teremos que pensar em alguma abordagem para os casos que o appstate não seja carregado. Tipo exibir um dialog informando ao usuário com um botão para tentar novamente ou uma abordagem do loading + tentativas automáticas de recarregar o load até pegar algo do servidor.

niltonvasques commented on app/src/androidTest/java/br/com/catbag/gifreduxsample/ui/giflist/GifListActivityTest.java:
```diff
@@ -313,53 +346,108 @@ public void whenListKeepOrderedAfterPlayTest() {
+    public void whenDefaultGifsAreLoaded() {
+        try {
+            DB db = DBFactory.open(getTargetContext());
+            db.destroy();
```
Which is the purpose of open and close database?

raulccabreu replied:
To destroy any traces of previous saved AppState on db.

raulccabreu commented:
The failure on build is some mock or AppState change leak between test cases.

raulccabreu committed:
- Use specifically persistence middleware replacing on  "whenDefaultGifsAreLoaded" test case to fix restMiddleware AppState leakage.

- Refactor the replace and clear middleware methods
- Dispatch empty AppState on "whenDefaultGifsAreLoaded" begin.
- Reorganize some methods place on TestHelper.

raulccabreu commented:
Fixed! :+1:

fpbitencourt commented on app/src/androidTest/java/br/com/catbag/gifreduxsample/ui/giflist/GifListActivityTest.java:
```diff
@@ -364,11 +362,14 @@ public void whenDefaultGifsAreLoaded() {
         try {
             DB db = DBFactory.open(getTargetContext());
             db.destroy();
+            db.close();
```
Faz mais sentido dar o close antes do destroy ou deixar somente o destroy caso o snappy garanta que feche o banco.

raulccabreu replied:
Pesquisando aqui, nos testes do SnappyDB o destroy tá sempre depois do close, e não é necessário ter o close antes, eles usam o destroy sem close tbm.

raulccabreu replied:
Removi o close e comitei.

fpbitencourt commented on app/src/androidTest/java/br/com/catbag/gifreduxsample/ui/giflist/GifListActivityTest.java:
```diff
@@ -364,11 +362,14 @@ public void whenDefaultGifsAreLoaded() {
-        mHelper.replaceMiddlewares(new DataManager(getTargetContext()), getTargetContext());
+        mHelper.dispatchFakeAppState(ImmutableAppState.builder().build());
+
+        mHelper.replacePersistenceMiddleware(new DataManager(getTargetContext()));
```
A definição dos middlewares não deveria vir antes de um dispatch?

raulccabreu replied:
Nesse caso ai é melhor que esteja antes, porque esse dispatch ai é só aquele que envia pro fake reducer e "injeta" um estado inicial no Fluxxan, nesse caso um estado vazio só pra garantir que nesse caso de teste o app funcione como se fosse usado pela primeira vez. Então eu não queria que o middleware pegasse essa mudança de estado pra fazer algo inesperado.

raulccabreu committed:
Removing unecessary snappy db.close() from whenDefaultGifsAreLoaded

fpbitencourt reviewed (approved):
:+1:
